### DISCUSSION TITLE
Collision sensor on a rigged object does not work

### DISCUSSION THREAD

**Discussion by wetoo-cando** [UPBGE/upbge#1795] [Announcements & others]
Hello,

I have a rigged hand object, for which I have set the Game Physics type to "Sensor". I have a collision sensor on this object inside logic bricks. When I press "Play", a grasping animation is triggered on the hand. The fingers then rotate and penetrate an icosphere near the palm. However, no collision with the sphere is detected.

Can someone please take a look. A zip file with the blend file inside is attached at the bottom (github does not allow directly attaching blend files to this discussion).

![image](https://user-images.githubusercontent.com/31558837/219348863-2c19d516-adf0-4731-9964-120b6cb7f55c.png)

The blend file is here:
[hands_low_poly_rig2.zip](https://github.com/UPBGE/upbge/files/10755213/hands_low_poly_rig2.zip)

**Comment by wetoo-cando** [comment]
I changed my approach according to: https://upbge.org/#/documentation/docs/latest/manual/manual/physics/introduction.html#mesh-deformations

So now I have
* A collision mesh (just a plane as the exact shape does not matter) -- this is the grandparent (parent of the armature) -- physics type: dynamic
* An Armature that drives the hand mesh -- parent of the hand mesh below -- physics type: "no collision" 
* A hand mesh (which is the display mesh whose animation/deformation is controlled by the Armature) -- physics type: "no collision"

I have a python component on the collision mesh, where in the `update()` method the `reinstancePhysicsMesh()` function is called with the hand mesh as argument. This should update the collision mesh to the deformed hand mesh (at least to my understanding). 

However, as you can see in the picture below, the collision mesh gets updated to some sort of icosphere (top right in the picture, inside the red bordered cube). I haven't been able to determine why. 

Can someone help?

Blend file: 
[hands_low_poly_rig2.zip](https://github.com/UPBGE/upbge/files/10757901/hands_low_poly_rig2.zip)

![image](https://user-images.githubusercontent.com/31558837/219420060-5e533f7a-71aa-4a5d-b729-ad89c4e8bfda.png)

**Comment by BluePrintRandom** [comment]
I would change your approach,

unparent and update the position and rot of child shapes relative to the
bones, and reparwnt them to a compound root object

**Reply by wetoo-cando** [reply]
@BluePrintRandom Thanks. Can you please elaborate?

If I understand you right:
- Rather than having Armature (parent)->hand_mesh(child) relationship, you suggest to have no such relationship
- "update the position and rot of child shapes relative to the bones": since constraints don't work inside UPBGE, I should copy the pos/rot from the bones to the mesh in python

"reparwnt them to a compound root object"  -- why/how would that help?

**Reply by wetoo-cando** [reply]
Thanks @BluePrintRandom 
I haven't been able to update the pose of child shapes relative to the bones. I tried this through object constraints, but these don't work inside the game engine. Also, the pos/rot values of the fingertip bone remain 0 even if the bone rotates at runtime. I don't know why.

> I would change your approach, unparent and update the position and rot of child shapes relative to the bones, and reparwnt them to a compound root object
> […](#)

**Comment by wetoo-cando** [comment]
Some progress. After unparenting the armature from the collision mesh, and correcting some scaling/offset issues, it seems the physics mesh for the hand is getting updated correctly as the hand grasps (deforms).

Collision detection of the fingers passing through the object still doesn't work though.

![hand-object](https://user-images.githubusercontent.com/31558837/219663325-3428666f-e92a-4058-9b1b-213ad6e83dc4.gif)

Blend file: 

[hands_low_poly_rig2_using_armature.zip](https://github.com/UPBGE/upbge/files/10767140/hands_low_poly_rig2_using_armature.zip)

**Comment by wetoo-cando** [comment]
Works finally:

![hand-object](https://user-images.githubusercontent.com/31558837/219698661-dfff5fd7-63e5-413d-b03c-0abfe78c9550.gif)

**Comment by BluePrintRandom** [comment]
If you parent a child shape to a compound root, it absorbs the child
collision shape to its own,

This is far more efficient than rebuilding triangle mesh physics.

**Reply by wetoo-cando** [reply]
Thanks!
